src/Microsoft.AspNetCore.ResponseCompression/BodyWrapperStream.cs

```diff
@@ -178,22 +178,33 @@ private void OnWrite()
             if (!_compressionChecked)
             {
                 _compressionChecked = true;
-
-                if (IsCompressable())
+                if (_provider.ShouldCompressResponse(_context))
                 {
-                    _response.Headers.Append(HeaderNames.ContentEncoding, _compressionProvider.EncodingName);
-                    _response.Headers.Remove(HeaderNames.ContentMD5); // Reset the MD5 because the content changed.
-                    _response.Headers.Remove(HeaderNames.ContentLength);
-
-                    _compressionStream = _compressionProvider.CreateStream(_bodyOriginalStream);
+                    var compressionProvider = CompressionProvider;
+                    if (compressionProvider != null)
+                    {
+                        _context.Response.Headers.Append(HeaderNames.ContentEncoding, compressionProvider.EncodingName);
+                        _context.Response.Headers.Remove(HeaderNames.ContentMD5); // Reset the MD5 because the content changed.
+                        _context.Response.Headers.Remove(HeaderNames.ContentLength);
+
+                        _compressionStream = compressionProvider.CreateStream(_bodyOriginalStream);
+                    }
                 }
             }
         }
 
-        private bool IsCompressable()
+        private ICompressionProvider CompressionProvider
```

Comment:
style: getters shouldn't have side-effects / hide functionality. make it a ResolveCompressionProvider method.